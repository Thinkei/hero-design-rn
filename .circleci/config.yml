restore_yarn_cache: &restore_yarn_cache
  restore_cache:
    key: yarn-v1-{{ checksum "yarn.lock" }}-{{ arch }}

save_yarn_cache: &save_yarn_cache
  save_cache:
    key: yarn-v1-{{ checksum "yarn.lock" }}-{{ arch }}
    paths:
      - ~/.cache/yarn

yarn_install: &yarn_install
  run:
    command: yarn install

default_config: &default_config
  working_directory: ~/app
  docker:
    - image: ehdevops/node:0.1.3
  environment:
    - CI: true

version: 2
jobs:
  lint:
    <<: *default_config
    steps:
      - checkout
      - setup_remote_docker
      - *restore_yarn_cache
      - *yarn_install
      - *save_yarn_cache
      - run: cd lib && yarn lint
  test:
    <<: *default_config
    steps:
      - checkout
      - setup_remote_docker
      - *restore_yarn_cache
      - *yarn_install
      - *save_yarn_cache
      - run: cd lib && yarn test
  release:
    <<: *default_config
    steps:
      - checkout
      - setup_remote_docker
      - *restore_yarn_cache
      - *yarn_install
      - *save_yarn_cache
      - attach_workspace:
        at: ~/app
      - run:
          name: Authenticate with registry
          command: echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > ~/app/.npmrc
      - run:
          name: Publish package
          command: cd lib && npm publish
workflows:
  version: 2
  build_and_deploy:
    jobs:
      - lint:
          context: org-global
      - test:
          context: org-global
      - release_lib:
          context: org-global
          filters:
            branches:
              only:
                - master
          requires:
            - lint
            - test
            - hold_release
      - hold_release:
          type: approval
          filters:
            branches:
              only: master

