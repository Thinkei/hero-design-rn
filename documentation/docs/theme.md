```md
{
  text: {
    fontWeight: '400',
    fontFamily: Font.default,
    defaultColor: BLACK,
    primaryColor: PRIMARY_LIGHT_GREY,
    secondaryColor: WHITE,
    dangerColor: RED,
    warningColor: PRIMARY_ORANGE,
    successColor: PRIMARY_BLUE,
    h1Style: {
      fontSize: normalizeText(22),
    },
    h2Style: {
      fontSize: normalizeText(20),
    },
    h3Style: {
      fontSize: normalizeText(18),
    },
    h4Style: {
      fontSize: normalizeText(16),
    },
    h5Style: {
      fontSize: normalizeText(14),
    },
    h6Style: {
      fontSize: normalizeText(12),
    },
    boldStyle: {
      fontWeight: '500',
    },
    bolderStyle: {
      fontWeight: '600',
    },
    lightStyle: {
      fontWeight: '300',
    },
    lighterStyle: {
      fontWeight: '200',
    },
  },
  colors: {
    primaryColor: PRIMARY_BLUE,
    dangerColor: RED,
    borderColor: PRIMARY_LIGHT_GREEN,
  },
  button: {
    textSize: 16,
    paddingVertical: 15,
  },
  input: {
    textSize: 16,
    paddingVertical: 15,
    paddingHorizontal: 15,
    marginBottom: 15,
    borderRadius: 4,
  },
}

```
