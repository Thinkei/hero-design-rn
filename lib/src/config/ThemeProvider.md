### Customized by theme
```md
import { ThemeProvider } from 'hero-design-rn'

const customeTheme = {
  text: {
    font: 'Arial'
  }
}
const App = () => {
  <ThemeProvider theme={customeTheme}>
    <Screen />
  <ThemeProvider/>
}
```

### Theme config
```md
{
  text: {
    fontFamily: string,
    fontWeight: string,
    defaultColor: string,
    primaryColor: string,
    secondaryColor: string,
    dangerColor: string,
    warningColor: string,
    successColor: string,
    h1Style: {
      fontSize: number
    },
    h2Style: {
      fontSize: number
    },
    h3Style: {
      fontSize: number
    },
    h4Style: {
      fontSize: number
    },
    h5Style: {
      fontSize: number
    },
    h6Style: {
      fontSize: number
    },
    boldStyle: {
      fontWeight: string
    },
    bolderStyle: {
      fontWeight: string
    },
    lightStyle: {
      fontWeight: string
    },
    lighterStyle: {
      fontWeight: string
    }
  },
  colors: {
    primaryColor: string,
    dangerColor: string,
    borderColor: string
  },
  button: {
    textSize: number,
    paddingVertical: number
  },
  input: {
    textSize: number,
    paddingVertical: number,
    paddingHorizontal: number,
    marginBottom: number,
    borderRadius: number
  }
};
```
