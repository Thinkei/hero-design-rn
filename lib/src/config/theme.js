// @flow

import normalizeText from "../helpers/normalizeText";

const theme: ThemeType = {
  text: {
    fontFamily: "Arial",
    fontWeight: "400",
    defaultColor: "#000000",
    primaryColor: "#000000",
    secondaryColor: "#ffffff",
    dangerColor: "red",
    warningColor: "yellow",
    successColor: "green",
    h1Style: {
      fontSize: normalizeText(25)
    },
    h2Style: {
      fontSize: normalizeText(20)
    },
    h3Style: {
      fontSize: normalizeText(18)
    },
    h4Style: {
      fontSize: normalizeText(15)
    },
    h5Style: {
      fontSize: normalizeText(12)
    },
    h6Style: {
      fontSize: normalizeText(10)
    },
    boldStyle: {
      fontWeight: "500"
    },
    bolderStyle: {
      fontWeight: "600"
    },
    lightStyle: {
      fontWeight: "300"
    },
    lighterStyle: {
      fontWeight: "200"
    }
  },
  colors: {
    primaryColor: "#003580",
    dangerColor: "#ff4136",
    borderColor: "#aaaaaa"
  },
  button: {
    textSize: normalizeText(16),
    paddingVertical: 15
  },
  input: {
    textSize: normalizeText(16),
    paddingVertical: 15,
    paddingHorizontal: 15,
    marginBottom: 15,
    borderRadius: 4
  }
};

export type ThemeType = {
  text: {
    fontFamily: string,
    fontWeight: string,
    defaultColor: string,
    primaryColor: string,
    secondaryColor: string,
    dangerColor: string,
    warningColor: string,
    successColor: string,
    h1Style: {
      fontSize: number
    },
    h2Style: {
      fontSize: number
    },
    h3Style: {
      fontSize: number
    },
    h4Style: {
      fontSize: number
    },
    h5Style: {
      fontSize: number
    },
    h6Style: {
      fontSize: number
    },
    boldStyle: {
      fontWeight: string
    },
    bolderStyle: {
      fontWeight: string
    },
    lightStyle: {
      fontWeight: string
    },
    lighterStyle: {
      fontWeight: string
    }
  },
  colors: {
    primaryColor: string,
    dangerColor: string,
    borderColor: string
  },
  button: {
    textSize: number,
    paddingVertical: number
  },
  input: {
    textSize: number,
    paddingVertical: number,
    paddingHorizontal: number,
    marginBottom: number,
    borderRadius: number
  }
};

let defaultTheme: ThemeType = theme;

const injectTheme = (theme: ThemeType): void => {
  defaultTheme = theme;
};
const getTheme = (): ThemeType => defaultTheme;

export { injectTheme, getTheme };
