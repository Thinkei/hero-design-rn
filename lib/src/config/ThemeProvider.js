// @flow
// this is from https://github.com/react-native-training/react-native-elements/blob/master/src/config/ThemeProvider.js
import React from "react";
import deepmerge from "deepmerge";

import theme, { type ThemeType } from "./theme";

const ThemeContext = React.createContext();

type ContextProps = { theme: ThemeType };
type ContextState = { theme: ThemeType };
export default class ThemeProvider extends React.Component<
  ContextProps,
  ContextState
> {
  constructor(props: ContextProps) {
    super(props);

    this.state = {
      theme: deepmerge(props.theme)
    };
  }

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.state.theme
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export const ThemeConsumer = ThemeContext.Consumer;
