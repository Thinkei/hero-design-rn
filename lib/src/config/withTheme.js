// @flow
// this is from https://github.com/react-native-training/react-native-elements/blob/master/src/config/withTheme.js

import * as React from "react";
import hoistNonReactStatics from "hoist-non-react-statics";

import { ThemeConsumer } from "./ThemeProvider";
import { getTheme } from "./theme";

const isClassComponent = Component =>
  Boolean(Component.prototype && Component.prototype.isReactComponent);

const withTheme = (WrappedComponent: React.ComponentType) => {
  class ThemedComponent extends React.Component {
    render() {
      const { forwardedRef, children, ...rest } = this.props;

      return (
        <ThemeConsumer>
          {context => {
            // If user isn't using ThemeProvider
            if (!context) {
              const props = { ...rest, theme: getTheme(), children };

              return isClassComponent(WrappedComponent) ? (
                <WrappedComponent ref={forwardedRef} {...props} />
              ) : (
                <WrappedComponent {...props} />
              );
            }

            const { theme } = context;
            const props = {
              theme,
              children,
              ...rest
            };

            if (isClassComponent(WrappedComponent)) {
              return <WrappedComponent ref={forwardedRef} {...props} />;
            }
            return <WrappedComponent {...props} />;
          }}
        </ThemeConsumer>
      );
    }
  }

  const name = `Themed.${WrappedComponent.displayName ||
    WrappedComponent.name ||
    "Component"}`;

  if (isClassComponent(WrappedComponent)) {
    const forwardRef = (props, ref) => (
      <ThemedComponent {...props} forwardedRef={ref} />
    );

    forwardRef.displayName = name;
    return hoistNonReactStatics(React.forwardRef(forwardRef), WrappedComponent);
  }

  ThemedComponent.displayName = name;
  return ThemedComponent;
};

export default withTheme;
