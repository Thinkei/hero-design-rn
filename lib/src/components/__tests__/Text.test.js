// @flow
import React from "react";

import { shallow } from "enzyme";

import ThemeProvider from "../../config/ThemeProvider";
import TextThemed, { Text } from "../Text";
import { getTheme } from "../../config/theme";
import renderer from "react-test-renderer";

const theme = getTheme();

describe("Text Component", () => {
  it("should render without issues", () => {
    const component = shallow(<Text theme={theme} />);

    expect(component.length).toBe(1);
    expect(component).toMatchSnapshot();
  });

  it("should have correct fontsize when h1", () => {
    const component = shallow(<Text h1 theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct fontsize when h2", () => {
    const component = shallow(<Text h2 theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct fontsize when h3", () => {
    const component = shallow(<Text h3 theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct fontsize when h4", () => {
    const component = shallow(<Text h4 theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct fontsize when h5", () => {
    const component = shallow(<Text h5 theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct color when primary", () => {
    const component = shallow(<Text primary theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct color when secondary", () => {
    const component = shallow(<Text secondary theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct color when danger", () => {
    const component = shallow(<Text danger theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct color when warning", () => {
    const component = shallow(<Text warning theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have correct color when success", () => {
    const component = shallow(<Text success theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it("should have text as children", () => {
    const component = shallow(<Text theme={theme}>Children Text</Text>);

    expect(component.props().children).toBe("Children Text");
  });

  it("should apply style prop as an array", () => {
    const component = shallow(
      <Text theme={theme} style={[{ color: "red" }, { fontSize: 30 }]}>
        Children Text
      </Text>
    );
    expect(component.props().style).toEqual({
      color: "red",
      fontSize: 30,
      fontFamily: "Arial",
      fontWeight: "400"
    });
  });

  it("should use values from the theme", () => {
    const theme = {
      text: {
        fontFamily: "Arial",
        fontWeight: "500",
        primaryColor: "#ffffff",
        h1Style: {
          fontSize: 50
        }
      }
    };

    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <TextThemed h1>Hey</TextThemed>
      </ThemeProvider>
    );
    expect(
      component.root.findByType(TextThemed).children[0].children[0].props.style
        .fontSize
    ).toEqual(50);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("local props should override style props on theme", () => {
    const theme = {
      text: {
        fontFamily: "Arial",
        fontWeight: "500",
        primaryColor: "#ffffff",
        h1Style: {
          fontSize: 50
        }
      }
    };

    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <TextThemed style={{ fontSize: 20 }}>Hey</TextThemed>
      </ThemeProvider>
    );

    expect(
      component.root.findByType(TextThemed).children[0].children[0].props.style
        .fontSize
    ).toBe(20);
    expect(component).toMatchSnapshot();
  });
});
