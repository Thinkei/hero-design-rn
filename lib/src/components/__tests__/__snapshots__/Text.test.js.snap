// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component local props should override style props on theme 1`] = `
<Text
  style={
    Object {
      "color": "#ffffff",
      "fontFamily": "Arial",
      "fontSize": 20,
      "fontWeight": "500",
    }
  }
>
  Hey
</Text>
`;

exports[`Text Component should have correct color when error 1`] = `
<Component
  style={
    Object {
      "color": "red",
      "fontFamily": "Arial",
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct color when primary 1`] = `
<Component
  style={
    Object {
      "color": "#000000",
      "fontFamily": "Arial",
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct color when secondary 1`] = `
<Component
  style={
    Object {
      "color": "#ffffff",
      "fontFamily": "Arial",
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct color when success 1`] = `
<Component
  style={
    Object {
      "color": "green",
      "fontFamily": "Arial",
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct fontsize when h1 1`] = `
<Component
  style={
    Object {
      "color": "#000000",
      "fontFamily": "Arial",
      "fontSize": 31.25,
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct fontsize when h2 1`] = `
<Component
  style={
    Object {
      "color": "#000000",
      "fontFamily": "Arial",
      "fontSize": 25,
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct fontsize when h3 1`] = `
<Component
  style={
    Object {
      "color": "#000000",
      "fontFamily": "Arial",
      "fontSize": 22.5,
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct fontsize when h4 1`] = `
<Component
  style={
    Object {
      "color": "#000000",
      "fontFamily": "Arial",
      "fontSize": 18.75,
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should have correct fontsize when h5 1`] = `
<Component
  style={
    Object {
      "color": "#000000",
      "fontFamily": "Arial",
      "fontSize": 15,
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should render without issues 1`] = `
<Component
  style={
    Object {
      "color": "#000000",
      "fontFamily": "Arial",
      "fontWeight": "400",
    }
  }
/>
`;

exports[`Text Component should use values from the theme 1`] = `
<Text
  style={
    Object {
      "color": "#ffffff",
      "fontFamily": "Arial",
      "fontSize": 50,
      "fontWeight": "500",
    }
  }
>
  Hey
</Text>
`;
